lasagne/layers/shape.py

```diff
@@ -300,3 +302,64 @@ def get_output_for(self, input, **kwargs):
         return padding.pad(input, self.width, self.val, self.batch_ndim)
 
 pad = PadLayer  # shortcut
+
+
+class SliceLayer(Layer):
+    """
+    Slices the input at a specific axis and at specific indices.
+
+    Parameters
+    ----------
+    incoming : a :class:`Layer` instance or a tuple
+        The layer feeding into this layer, or the expected input shape
+
+    indices : int, list, tuple or slice instance
+        When indices is an int the corresponding index is selected from the
+        axis dimension and the axis dimension is dropped. When indices is list,
+        tuple or slice instance the axis dimension is kept even if it is
+        singleton dimension after slicing. Negative values behaves as in numpy
+        indexing.
+
+    axis : int
+        Specifies the axis from which the indices are selected. Negative values
+        behaves as in numpy indexing.
+
+    Examples
+    --------
+    >>> from lasagne.layers import SliceLayer, InputLayer, get_output
+    >>> in_shp = (2, 3, 2)
+    >>> l_inp = InputLayer(in_shp)
+    >>> l_slice_ax0 = SliceLayer(l_inp, axis=0, indices=0)
+    >>> l_slice_ax1 = SliceLayer(l_inp, axis=1, indices=(1, 3))
+    >>> l_slice_ax2 = SliceLayer(l_inp, axis=-1, indices=-1)
+    >>> x = np.arange(np.prod(in_shp)).reshape(in_shp).astype('float32')
+    >>> get_output(l_slice_ax0, x).eval().shape
+    (3, 2)
+    >>> get_output(l_slice_ax1, x).eval().shape
+    (2, 2, 2)
+    >>> get_output(l_slice_ax2, x).eval().shape
+    (2, 3)
```

Comment:
I'd simplify that a bit, and also reorder it so the output shape comes directly after the layer definition (similarly to how I did it for the ReshapeLayer). Also I'd use three different sizes for the three dimensions to make it clearer.
```
>>> l_in = InputLayer((2, 3, 4))
>>> SliceLayer(l_in, indices=0, axis=0).output_shape
... # equals input[0]
(3, 4)
>>> SliceLayer(l_in, indices=slice(1, 3), axis=1).output_shape
... # equals input[:, 1:3]
(2, 2, 4)
>>> SliceLayer(l_in, indices=slice(-2, -1), axis=-1)
... # equals input[..., -2:-1]
2, 3, 1
```
This also illustrates that a slice of only one element will still keep the axis. What do you think?